From 49373fd01465d56146fa5f1ceded3907ad520ca0 Mon Sep 17 00:00:00 2001
From: Lei Zhu <julytreee@gmail.com>
Date: Sun, 14 Aug 2022 22:44:59 +0800
Subject: [PATCH] Fix #3047 (#3092)

* Make path canonicalized

* Update extra-source-files

* Replace with normalise

* Change to a more detailed log

* Comment patch detail

* 2 spaces indent
---
 exe/Plugins.hs                                |   2 +-
 .../hls-module-name-plugin.cabal              |   2 +
 .../src/Ide/Plugin/ModuleName.hs              | 140 +++++++++++-------
 plugins/hls-module-name-plugin/test/Main.hs   |  15 +-
 .../test/testdata/cabal.project               |   1 +
 .../testdata/canonicalize/Lib/A.expected.hs   |   1 +
 .../test/testdata/canonicalize/Lib/A.hs       |   0
 .../testdata/canonicalize/canonicalize.cabal  |   7 +
 .../test/testdata/hie.yaml                    |  23 ++-
 9 files changed, 124 insertions(+), 67 deletions(-)
 create mode 100644 plugins/hls-module-name-plugin/test/testdata/cabal.project
 create mode 100644 plugins/hls-module-name-plugin/test/testdata/canonicalize/Lib/A.expected.hs
 create mode 100644 plugins/hls-module-name-plugin/test/testdata/canonicalize/Lib/A.hs
 create mode 100644 plugins/hls-module-name-plugin/test/testdata/canonicalize/canonicalize.cabal

diff --git a/plugins/hls-module-name-plugin/hls-module-name-plugin.cabal b/plugins/hls-module-name-plugin/hls-module-name-plugin.cabal
index 29fbe521..1290ab75 100644
--- a/plugins/hls-module-name-plugin/hls-module-name-plugin.cabal
+++ b/plugins/hls-module-name-plugin/hls-module-name-plugin.cabal
@@ -16,6 +16,8 @@ extra-source-files:
   LICENSE
   test/testdata/**/*.yaml
   test/testdata/**/*.hs
+  test/testdata/**/*.cabal
+  test/testdata/**/*.project
 
 library
   exposed-modules:  Ide.Plugin.ModuleName
diff --git a/plugins/hls-module-name-plugin/test/Main.hs b/plugins/hls-module-name-plugin/test/Main.hs
index ce0fa1e7..914fcb69 100644
--- a/plugins/hls-module-name-plugin/test/Main.hs
+++ b/plugins/hls-module-name-plugin/test/Main.hs
@@ -39,10 +39,15 @@ tests =
       void $ skipManyTill anyMessage (message SWorkspaceApplyEdit)
   , testCase "Should not show code lens if the module name is correct" $
       runSessionWithServer moduleNamePlugin testDataDir $ do
-          doc <- openDoc "CorrectName.hs" "haskell"
-          lenses <- getCodeLenses doc
-          liftIO $ lenses @?= []
-          closeDoc doc
+        doc <- openDoc "CorrectName.hs" "haskell"
+        lenses <- getCodeLenses doc
+        liftIO $ lenses @?= []
+        closeDoc doc
+  -- https://github.com/haskell/haskell-language-server/issues/3047
+  , goldenWithModuleName "Fix#3047" "canonicalize/Lib/A" $ \doc -> do
+      [CodeLens { _command = Just c }] <- getCodeLenses doc
+      executeCommand c
+      void $ skipManyTill anyMessage (message SWorkspaceApplyEdit)
   ]
 
 goldenWithModuleName :: TestName -> FilePath -> (TextDocumentIdentifier -> Session ()) -> TestTree
diff --git a/plugins/hls-module-name-plugin/test/testdata/cabal.project b/plugins/hls-module-name-plugin/test/testdata/cabal.project
new file mode 100644
index 00000000..1406cd09
--- /dev/null
+++ b/plugins/hls-module-name-plugin/test/testdata/cabal.project
@@ -0,0 +1 @@
+packages: ./canonicalize
diff --git a/plugins/hls-module-name-plugin/test/testdata/canonicalize/Lib/A.expected.hs b/plugins/hls-module-name-plugin/test/testdata/canonicalize/Lib/A.expected.hs
new file mode 100644
index 00000000..c5877f71
--- /dev/null
+++ b/plugins/hls-module-name-plugin/test/testdata/canonicalize/Lib/A.expected.hs
@@ -0,0 +1 @@
+module Lib.A where
diff --git a/plugins/hls-module-name-plugin/test/testdata/canonicalize/Lib/A.hs b/plugins/hls-module-name-plugin/test/testdata/canonicalize/Lib/A.hs
new file mode 100644
index 00000000..e69de29b
diff --git a/plugins/hls-module-name-plugin/test/testdata/canonicalize/canonicalize.cabal b/plugins/hls-module-name-plugin/test/testdata/canonicalize/canonicalize.cabal
new file mode 100644
index 00000000..dc0e099e
--- /dev/null
+++ b/plugins/hls-module-name-plugin/test/testdata/canonicalize/canonicalize.cabal
@@ -0,0 +1,7 @@
+cabal-version: 2.4
+name: canonicalize
+version: 0.1.0.0
+
+library
+  build-depends: base
+  hs-source-dirs: ./
diff --git a/plugins/hls-module-name-plugin/test/testdata/hie.yaml b/plugins/hls-module-name-plugin/test/testdata/hie.yaml
index 022fee55..fb1c7521 100644
--- a/plugins/hls-module-name-plugin/test/testdata/hie.yaml
+++ b/plugins/hls-module-name-plugin/test/testdata/hie.yaml
@@ -1,8 +1,17 @@
 cradle:
-  direct:
-    arguments:
-      - "-isubdir"
-      - "TEmptyModule"
-      - "TWrongModuleName"
-      - "mainlike"
-      - "CorrectName"
+  multi:
+    - path: "./"
+      config:
+        cradle:
+          direct:
+            arguments:
+              - "-isubdir"
+              - "TEmptyModule"
+              - "TWrongModuleName"
+              - "CorrectName"
+    - path: "./canonicalize"
+      config:
+        cradle:
+          cabal:
+            - path: "./"
+              component: "lib:canonicalize"
-- 
2.25.1

